extensions/README.md

```diff
@@ -0,0 +1,362 @@
+the value of the indicated type, along with possible errors. ("Must" versions
+of these also exists, e.g. `MustGetString()`, that do not return an error).
+
+You can also get the pointer your extensions instance itself by calling
+extended type instance's `GetExtensions()`. This returns an `interface{}`, so
+you will need to type assert in order to be able to access the fields directly.
+
+### Introducing additional constraints
+
+To introduce new constraints,  add a method called `Validate<type>(v *<type>)`
+to your extensions struct, where `<type>` is the name of the type being
+extended (one of the ones listed above) -- e.g.
+`ValidateComid(v *comid.Comid)` when extending `comid.Comid`. This method, if
+it exists, will be invoked inside the extended type instance's `Valid()`
+method, passing itself as the parameter.
+
+You do not need to define this method, unless you actually want to enforce some
+constraints (i.e. if you just want to define additional fields).
+
+### Example
+
+The following example illustrates how to implement a map extension by extending `comid.Entity` with the following features:
+
+1. an optional "email" field
+2. additional validation to ensure that the existing name field contains a
+   valid UUID (note: since `NameEntry` is a type choice extensible, this can
+   also be done by defining a new value type for `NameEntry` -- see the
+   following section).
+
+```go
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+
+	"github.com/google/uuid"
+	"github.com/veraison/corim/comid"
+)
+
+type EntityExtensions struct {
+	Email string `cbor:"-1,keyasint,omitempty" json:"email,omitempty"`
+}
+
+func (o EntityExtensions) ValidEntity(val *comid.Entity) error {
+	_, err := uuid.Parse(val.EntityName.String())
+	if err != nil {
+		return fmt.Errorf("invalid UUID: %w", err)
+	}
+
+	return nil
+}
+
+var sampleText = `
+{
+      "name": "31fb5abf-023e-4992-aa4e-95f9c1503bfa",
+      "regid": "https://acme.example",
+      "email": "info@acme.com",
+      "roles": [
+        "tagCreator",
+        "creator",
+        "maintainer"
+      ]
+}
+`
+
+func main() {
+	var entity comid.Entity
+	entity.RegisterExtensions(&EntityExtensions{})
+
+	if err := json.Unmarshal([]byte(sampleText), &entity); err != nil {
+		log.Fatalf("ERROR: %s", err.Error())
+	}
+
+	if err := entity.Valid(); err != nil {
+		log.Fatalf("failed to validate: %s", err.Error())
+	} else {
+		fmt.Println("validation succeeded")
+	}
+
+    // obtain the extension field value via a generic getter
+	email := entity.Extensions.MustGetString("email")
+	fmt.Printf("entity email: %s\n", email)
+
+    // retrive the extensions struct and get value via its field.
+	exts := entity.GetExtensions().(*EntityExtensions)
+	fmt.Printf("also entity email: %s\n", exts.Email)
+}
+```
+
+
+## Type Choice Extensions
+
+Type Choice extensions allow specifying alternative types for existing CoRIM
+fields by defining a struct that implements an appropriate interface and
```

Comment:
```suggestion
fields by defining a  type that implements an appropriate interface and
```

Reply:
I am confused, I thought for Type extensions we will register the new type and a Callback which instantiates the new Type.

Why we are requesting users to create a structure?

Reply:
- "struct" rather than "type" is correct here, since we're talking about concrete implementations (the interface being implemented is the type).
  - I assume, by "callback" you mean "factory"? The factory still needs something to actually put into the new instance that that something is the struct that needs to be implemented. E.g. need to provide an implementation for both `IClassIDValue` which is what you're actually implementing, and a `IClassIDFactory` that produces a `*ClassID` that is populated with the underlying value of your type.

Reply:
> * "struct" rather than "type" is correct here, since we're talking about concrete implementations (the interface being implemented is the type).
> * I assume, by "callback" you mean "factory"? The factory still needs something to actually put into the new instance that that something is the struct that needs to be implemented. E.g. need to provide an implementation for both `IClassIDValue` which is what you're actually implementing, and a `IClassIDFactory` that produces a `*ClassID` that is populated with the underlying value of your type.

As discussed the user can define a type like taggedInt and provide the required methods for instantiation.

I think, we need to refine the language, to make it super clear to the consumer of the extensions!